File: algorithms/2_medium/S/01574_shortest_subarray_to_be_removed_to_make_array_sorted/solution.c
typedef ptrdiff_t isize;

static inline isize custom_longest(const int* sorted_prefix, const isize sorted_prefix_len, 
                                   const int* sorted_suffix, const isize sorted_suffix_len)
{
    int* sub = malloc((size_t)sorted_prefix_len * sizeof(*sub));
    assert(sub != NULL);
    for (isize i = 0; i < sorted_prefix_len; i += 1) { sub[i] = sorted_prefix[i]; }
    const isize sub_len = sorted_prefix_len;
    isize res = sub_len;

    for (isize i = 0; i < sorted_suffix_len; i += 1) {
        const int a = sorted_suffix[i];
        isize hi = sub_len - 1;
        isize lo = 0;

        while (lo <= hi) {
            const isize mid = lo + (hi - lo) / 2;

            if (sub[mid] <= a) {
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }

        if (lo == sub_len) {
            // all remaining elems are sorted and can be added to current sub
            res = (sorted_suffix_len - i) + sub_len;
            goto cleanup;
        } else {
            sub[lo] = a;
        }
    }

cleanup:
    free(sub);
    return res;
}

int findLengthOfShortestSubarray(int* arr, int arr_size)
{
    assert(arr != NULL && arr_size >= 0);
    const isize len = arr_size;
    if (len <= 1) { return 0; }
    isize hi = len - 1;
    isize lo = 0;

    while (hi > 0 && arr[hi - 1] <= arr[hi]) {
        hi -= 1;
    }

    if (hi == 0) {
        return 0;
    }

    while (arr[lo] <= arr[lo + 1]) {
        lo += 1;
    }

    const int* sorted_prefix = &arr[0];
    const isize sorted_prefix_len = lo + 1;
    const int* sorted_suffix = &arr[hi];
    const isize sorted_suffix_len = len - hi;

    const isize longest = custom_longest(sorted_prefix, sorted_prefix_len,
                                       sorted_suffix, sorted_suffix_len);
    return (int)(len - longest);
}

File: algorithms/2_medium/S/01574_shortest_subarray_to_be_removed_to_make_array_sorted/solution.rs
impl Solution {
    #[inline(always)]
    fn custom_longest(sorted_prefix: &[i32], sorted_suffix: &[i32]) -> usize {
        let mut sub = sorted_prefix.to_vec().into_boxed_slice();
        let sub_len = sub.len();
        let sorted_suffix_len = sorted_suffix.len();

        for (i, &a) in sorted_suffix.into_iter().enumerate() {
            let mut hi = sub_len as isize - 1;
            let mut lo = 0_isize;

            while lo <= hi {
                let mid = lo + (hi - lo) / 2;

                if sub[mid as usize] <= a {
                    lo = mid + 1;
                } else {
                    hi = mid - 1;
                }
            }

            if lo as usize == sub_len {
                // all remaining elems are sorted and can be added to current sub
                return (sorted_suffix_len - i) + sub_len;
            } else {
                sub[lo as usize] = a;
            }
        }

        sub_len
    }

    pub fn find_length_of_shortest_subarray(arr: Vec<i32>) -> i32 {
        let len = arr.len();
        if len <= 1 { return 0; }
        let mut hi = len - 1;
        let mut lo = 0_usize;

        while hi > 0 && arr[hi - 1] <= arr[hi] {
            hi -= 1;
        }

        if hi == 0 {
            return 0;
        }

        while arr[lo] <= arr[lo + 1] {
            lo += 1;
        }

        let sorted_prefix = &arr[0..=lo];
        let sorted_suffix = &arr[hi..];

        (len - Self::custom_longest(sorted_prefix, sorted_suffix)).try_into().unwrap()
    }
}
